pkg/protocol/faststate/faststate.go

```diff
@@ -0,0 +1,129 @@
+// Package faststate contains a generic state machine implementation that is
```

Comment:
I'm wondering about alternative ways of structuring the code of both state machines. Packages `state` and `faststate` can be really mysterious for someone not familiar with this code and the differences are not obvious at the first sight.

What comes into my mind is the fact that we can call the `state.Machine` a "synchronous" machine as states are synchronized using the block counter. On the other hand, the new `faststate.Machine` is "asynchronous" as states decide about the transition using their own internal predicates. That said we can refactor the current structure twofold:

**Option 1**: We can rename the package `state` to `syncstate` and `faststate` to `asyncstate`. This way we will end up with the following types:
- `syncstate.State`
- `syncstate.Machine`
- `asyncstate.State`
- `asyncstate.Machine`

**Option 2**: We can move everything into a single package `state` and rename specific types. This way we would have:
- `state.SyncState`
- `state.SyncMachine`
- `state.AsyncState`
- `state.AsyncMachine`

Although **O1** can be tempting as we would have two separate and small packages with well-defined types, I think we should lean towards **O2**. The main reasons are:
- All this code is about state and both machines (and associated types) have a lot of common things so it makes sense to keep them in a single `state` package. We can probably extract some common parts to reduce duplication (both `State` interfaces share a lot of methods).
- This option is closer to how we typically group things into packages. We deal with some different flavors of the same component. For example, we have the `pkg/tecdsa/retry` package that exposes two retry algorithms, for two separate use cases.
- The size of the resulting package will still be on a sane level

Reply:
Sure, let's do it separately!

Reply:
Done in https://github.com/keep-network/keep-core/pull/3365.